File: exercism/go/archive/twelve-days/twelve_days.go
// Package twelve is our very holiday-themed package to return a song
package twelve

import "bytes"

// testVersion represents the test suite we're up against
const testVersion = 1

/// songDay is the basic structure of each day in the song
type songDay struct {
	day  string
	gift string
}

// global var for us to use
var song = make(map[int]songDay, 12)

// initSong is an internal function to initialize our song
func initSong() bool {
	song[0] = songDay{day: "first", gift: "a Partridge in a Pear Tree."}
	song[1] = songDay{day: "second", gift: "two Turtle Doves"}
	song[2] = songDay{day: "third", gift: "three French Hens"}
	song[3] = songDay{day: "fourth", gift: "four Calling Birds"}
	song[4] = songDay{day: "fifth", gift: "five Gold Rings"}
	song[5] = songDay{day: "sixth", gift: "six Geese-a-Laying"}
	song[6] = songDay{day: "seventh", gift: "seven Swans-a-Swimming"}
	song[7] = songDay{day: "eighth", gift: "eight Maids-a-Milking"}
	song[8] = songDay{day: "ninth", gift: "nine Ladies Dancing"}
	song[9] = songDay{day: "tenth", gift: "ten Lords-a-Leaping"}
	song[10] = songDay{day: "eleventh", gift: "eleven Pipers Piping"}
	song[11] = songDay{day: "twelfth", gift: "twelve Drummers Drumming"}

	return true
}

// Song will return the entire song
func Song() string {
	initSong()

	var buf bytes.Buffer

	for x := 0; x < 12; x++ {
		buf.Write([]byte("On the "))
		buf.Write([]byte(song[x].day))
		buf.Write([]byte(" day of Christmas my true love gave to me, "))
		buf.Write([]byte(song[x].gift))
		for y := x - 1; y >= 0; y-- {
			buf.Write([]byte(", "))
			if y == 0 {
				buf.Write([]byte("and "))
				buf.Write([]byte(song[y].gift))
				break
			}

			buf.Write([]byte(song[y].gift))
		}
		buf.Write([]byte("\n"))
	}

	return buf.String()
}

// Verse will return a specific portion of the song
func Verse(verse int) string {
	initSong()

	verse = verse - 1 // off-by-one

	var buf bytes.Buffer

	buf.Write([]byte("On the "))
	buf.Write([]byte(song[verse].day))
	buf.Write([]byte(" day of Christmas my true love gave to me, "))
	buf.Write([]byte(song[verse].gift))
	for y := verse - 1; y >= 0; y-- {
		buf.Write([]byte(", "))
		if y == 0 {
			buf.Write([]byte("and "))
			buf.Write([]byte(song[y].gift))
			break
		}

		buf.Write([]byte(song[y].gift))
	}

	return buf.String()
}

File: exercism/python/twelve-days/twelve_days.py
"""Module to recite the Twelve Days of Christmas song."""


LYRICS = {
    1: ("first", "a Partridge in a Pear Tree"),
    2: ("second", "two Turtle Doves"),
    3: ("third", "three French Hens"),
    4: ("fourth", "four Calling Birds"),
    5: ("fifth", "five Gold Rings"),
    6: ("sixth", "six Geese-a-Laying"),
    7: ("seventh", "seven Swans-a-Swimming"),
    8: ("eighth", "eight Maids-a-Milking"),
    9: ("ninth", "nine Ladies Dancing"),
    10: ("tenth", "ten Lords-a-Leaping"),
    11: ("eleventh", "eleven Pipers Piping"),
    12: ("twelfth", "twelve Drummers Drumming"),
}


def recite(start_verse: int, end_verse: int) -> list[str]:
    """Generate the verses for the Twelve Days of Christmas song.

    :param start_verse: int - start verse number
    :param end_verse: int - end verse number
    :return: list[str] - series of verses for the song between `start_verse` and `end_verse`.
    """
    start_line = f"On the {LYRICS[start_verse][0]} day of Christmas my true love gave to me: "

    for verse in range(start_verse, 1, -1):
        start_line += f"{LYRICS[verse][1]}, "

    if start_verse != 1:
        start_line += "and "

    start_line += f"{LYRICS[1][1]}."
    if start_verse == end_verse:
        return [start_line]

    return [start_line] + recite(start_verse + 1, end_verse)
